runtime/queries/metricsview_comparison_toplist.go

```diff
@@ -99,13 +104,75 @@ func (q *MetricsViewComparison) Resolve(ctx context.Context, rt *runtime.Runtime
 		q.Where = convertFilterToExpression(q.Filter)
 	}
 
+	err = q.calculateMeasuresMeta()
+	if err != nil {
+		return err
+	}
+
 	if !isTimeRangeNil(q.ComparisonTimeRange) {
 		return q.executeComparisonToplist(ctx, olap, q.MetricsView, priority, q.ResolvedMVSecurity)
 	}
 
 	return q.executeToplist(ctx, olap, q.MetricsView, priority, q.ResolvedMVSecurity)
 }
 
+func (q *MetricsViewComparison) calculateMeasuresMeta() error {
+	measureMap := make(map[string]bool, len(q.Measures))
+	for _, m := range q.Measures {
+		measureMap[m.Name] = true
+	}
+
+	compare := !isTimeRangeNil(q.ComparisonTimeRange)
+
+	if !compare && len(q.ComparisonMeasures) > 0 {
+		return fmt.Errorf("comparison measures are provided but comparison time range is not")
+	}
+
+	comparisonMap := make(map[string]bool, len(q.ComparisonMeasures))
+
+	// backwards compatibility
+	if len(q.ComparisonMeasures) == 0 && compare {
+		comparisonMap = measureMap
+	}
+
+	for _, m := range q.ComparisonMeasures {
+		if !measureMap[m] {
+			return fmt.Errorf("comparison measure '%s' is not present in the metrics view '%s'", m, q.MetricsViewName)
+		}
+		comparisonMap[m] = true
+	}
+
+	err := validateSort(q.Sort, comparisonMap, compare)
+	if err != nil {
+		return err
+	}
+
+	err = validateAlias(q.Aliases, comparisonMap)
+	if err != nil {
+		return err
+	}
+
+	q.MeasuresMeta = make(map[string]metricsViewMeasureMeta, len(q.Measures))
+
+	inner := 1
+	outer := 1
+	for _, m := range q.Measures {
+		expand := comparisonMap[m.Name]
+		q.MeasuresMeta[m.Name] = metricsViewMeasureMeta{
+			innerIndex: inner,
+			outerIndex: outer,
+			expand:     expand,
+		}
+		if expand {
+			outer += 4
+		} else {
+			outer++
+		}
+		inner++
+	}
+	return nil
+}
```

Comment:
Always hard to say about premature optimization, but since this gets hit dozens of times per dashboard load, I think there are some low hanging optimizations that can be easily fixed here:

1. The number of measures is usually quite small, so instead of allocating/populating/garbage collecting maps for `measureMap` and `comparisonMap`, consider just iterating over the input slices when building `q.MeasureMeta`
2. Can populate and pass `q.MeasuresMeta` to `validateSort` and `validateAlias` instead of the separate maps
3. For the case where `len(q.ComparisonMeasures) == 0 && compare`, there is no need to validate the comparison measures

Another question – what about case sensitivity? I'm not sure, but I wonder if the old implementation was case insensitive?

Reply:
fixed